core/shared/src/main/scala/org/http4s/websocket/WebSocketFrameAggregator.scala

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2013 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.websocket
+
+import fs2.Chunk
+import fs2.Pipe
+import scodec.bits.ByteVector
+
+private[http4s] object WebSocketFrameAggregator {
+  def aggregateFragment[F[_]]: Pipe[F, WebSocketFrame, WebSocketFrame] =
+    stream => {
+      // Takes a chunk of WebSocketFrames and aggregates fragmented frames into one frame.
+      def aggregate(
+          frames: Chunk[WebSocketFrame]
+      ): (Chunk[WebSocketFrame], Chunk[WebSocketFrame]) = {
+        val initialState = (Chunk.empty[WebSocketFrame], Chunk.empty[WebSocketFrame])
+        frames.foldLeft(initialState) {
+          case ((fragments, result), curFrame) if curFrame.last =>
+            // Current frame is a single frame (not fragmented), or the last one of a sequence of fragments.
+            val fragmentSum = fragments ++ Chunk(curFrame)
+            val aggregatedData = fragmentSum.map(_.data).foldLeft(ByteVector.empty)(_ ++ _)
```

Comment:
if we do the `_.data` in the fold lambda we can save an intermediate allocation.

Reply:
done: [0000be3](https://github.com/http4s/http4s/pull/7091/commits/0000be3f7907c6018c295c62a2e155cdb374dd26)